api/solarperformanceinsight_api/compute.py

```diff
@@ -444,3 +501,182 @@ def compare_expected_and_actual(job: models.StoredJob, si: storage.StorageInterf
+        zip_longest(
+            chains,
+            ref_weather_gen,
+            actual_weather_gen,
+            ref_pac_gen,
+            ref_pdc_gen,
+        )
+    ):
+        # use the ModelChain for converting any temperature to cell temperature and
+        # converting irradiance to POA. In the future, could only run these conversions
+        # instead of full modelchain calculation.
+        # since it sets class properties, copy for the actuals calculations
+        chain_actual = deepcopy(chain)
+        pac0 = job.definition.system_definition.inverters[i].inverter_parameters._pac0
+        num_arrays = len(chain.system.arrays)
+        gammas = [
+            arr._gamma for arr in job.definition.system_definition.inverters[i].arrays
+        ]
+        if any([g is None for g in gammas]):
+            raise TypeError(
+                "Currently unable to compare predicted and actual performance for "
+                "PVsyst specified arrays."
+            )
+        if data_available == models.PredictedDataEnum.weather_only:  # 2A-4
+            db_results, _ = process_single_modelchain(
+                chain, ref_weather, ref_model_method, tshift, i
+            )
+            results_list += db_results
+            ref_pdc = adjust(_get_mc_dc(chain.results, num_arrays))  # type: ignore
+        else:
+            # run chain on ref weather
+            getattr(chain, ref_model_method)(
+                [d.shift(freq=tshift) for d in ref_weather]  # type: ignore
+            )
+
+        # run chain on actual weather
+        getattr(chain_actual, actual_model_method)(
+            [d.shift(freq=tshift) for d in actual_weather]  # type: ignore
+        )
+        # use pvlib.modelchain._irrad_for_celltemp that returns POA global if available
+        # otherwise uses effective irradiance
+        poa_ref = _irrad_for_celltemp(
+            chain.results.total_irrad, chain.results.effective_irradiance
+        )
+        poa_actual = _irrad_for_celltemp(
+            chain_actual.results.total_irrad, chain_actual.results.effective_irradiance
+        )
+        # use cell temperature from the pvlib modelchain
+        # If air temp + wind speed or module temperature were supplied, they are
+        # converted to cell temperature via the array's temperature model
+        t_ref = chain.results.cell_temperature
+        t_actual = chain_actual.results.cell_temperature
+
+        # mean of array POArat * TempFactor for this inverter
+        # could make more sense to use weighted mean with weights set
+        # by array power percentage
+        # another alternative, calculate average POArat and TempFactor separately
+        poa_rat = [pa / pr for pa, pr in zip(poa_actual, poa_ref)]
+        tempfactor = list(map(_temp_factor, gammas, t_ref, t_actual))
+        poa_rat_x_temp_factor = adjust(  # modelchain outputs are all shifted
+            sum([p * t for p, t in zip(poa_rat, tempfactor)]) / num_arrays
+        )
```

Comment:
@cwhanse is this OK or would another method be preferred? Right now, gamma could be different for each array, and we are allowing uploading the weather data for each array, so we need some way to aggregate to the inverter level where Pac/Pdc are defined.

Reply:
If I follow this code correctly: `poa_rat` and `tempfactor` are per array (per array derived from `poa_actual`. `poa_rat_x_temp_factor` is an average of the adjustment factor (averaged over arrays).

If I have that right, then I think this method is OK. The case where it could be improved is if a user can upload ref_pdc per array, but I don't think we envision that option. With per-array pdc, we would calculate `pdc_ref_adj` per array, sum the result then apply the inverter efficiency factor and clipping limit.

Reply:
Yes, you have that right, and yes, we don't plan on supporting per-array pdc.